JMESPath docs, unit test for jmes functions and merge with master (#1)
EchoShoot/parsel: docs/usage.rst

Usage
=====

> Create a :class:`~parsel.selector.Selector` object for the input text that you want to parse and use.

Reviewer: Please, keep the original line length limit of 79.

Also, not a strong opinion, but I don’t see why you felt the need to add “and use” at the end. I would rather simplify the line further instead:

```suggestion
Create a :class:`~parsel.selector.Selector` object for your input text.
```

Author: Those are from the original usage, I didnt put them there.
Edit: you meant only the "and use". Yeah I guess I put that somewhere after messing with JMES. Sorry.

> For HTML or XML, use `CSS`_ or `XPath`_ expressions to select elements::

Reviewer: I know this comes from the original string, but I think it was not technically accurate (e.g. text is not an element).

```suggestion
For HTML or XML, use `CSS`_ or `XPath`_ expressions to select data::
```

Author: I guess if you don't use get/getall we could say the selector is handling elements. But overall its true we generally just want the data inside those elements anyway.

    >>> from parsel import Selector
    >>> html_text = "<html><body><h1>Hello, Parsel!</h1></body></html>"
    >>> html_selector = Selector(text=html_text)
    >>> html_selector.css('h1')
    [<Selector xpath='descendant-or-self::h1' data='<h1>Hello, Parsel!</h1>'>]
    >>> html_selector.xpath('//h1')  # the same, but now with XPath
    [<Selector xpath='//h1' data='<h1>Hello, Parsel!</h1>'>]

> For JSON format text, use `JMESPath`_ expressions to select fields inside the text::

Reviewer: In line with the HTML/XML line above:

```suggestion
For JSON, use `JMESPath`_ expressions to select data::
```

    >>> json_text = '{"title":"Hello, Parsel!"}'
    >>> json_selector = Selector(text=json_text)
    >>> json_selector.jmespath('title')

>     [<Selector xpath='title' data='Hello, Parsel!'>]

Reviewer: Hmm… Is this the actual output? Maybe we need to fix this. While it is OK for CSS expressions (since they are converted to XPath), here we should probably print `jmespath='title'` instead.

Author: Yes, the selector displays xpath='title'. Tested now with an only-json selector.

Reviewer: How do you feel about looking into fixing that in code?

Author: I looked at it and was thinking of giving my assessment here over the __str__ of the Selector. 
It feels like the problem could be easily fixed by changing the` self.type `to json during a succesful `.jmespath` query. Nevertheless I think this shouldn't be the approach.
The proper way to do it seems to be guaranteeing the `self.type` setup will recognize json, but I stopped there, I felt maybe I was overstepping and would just hinder the progress of the docs.
So yeah if you are opening the possibility for me to fix it then yeah sure I will look further!

Reviewer: Maybe there is another way… What if we make it so that, instead of `xpath`, it says `query`?

Author: The "intended" way as per the tool would be `jmespath`, but I believe it can be `query` with no problems. 
The main issue is how to approach this change.
The simplest one I belive would be to, whenever jmespath runs and type is None, use _get_root with type JSON and then at the __str dunder we change `jmespath` to `query`

Reviewer: What I am suggesting is to change `xpath` to `query` in `__str__` for every case, regardless of the `type`.

Author: Oh, yeah in that case it simplifies a lot the code. Let's go with that approach!

And extract data from those elements::

    >>> html_selector.css('h1::text').get()
    'Hello, Parsel!'
    >>> html_selector.xpath('//h1/text()').getall()
    ['Hello, Parsel!']
    >>> json_selector.jmespath('title').get()

>     'Hello, Parsel!'

Reviewer: I have just realized that the goal here was not to feature both CSS and XPath, but to feature both `get` and `getall`. So you could either leave it as it was, or maybe better, remove the XPath one and leave the JMESPath one using  `getall`.

Author: I think its good to display one simple example for each selector, as it completes the first level of the "tutorial" without leaving anything to imagination.
Maybe we could have 3 examples of get, one for each selector, and after them, do another line saying that getall can be used to get multiple data, and then display an example that brings multiple data, such as links?

Reviewer: Maybe keep your change, and once we merge text support we can have one example here use it.

Reviewer: Wait, you took the  only `getall()` instead.

Author: I thought that it would be better to have all 3 examples retrieving the same thing, and briefly mention` getall()` as it feels that an explanation of it will suit it better than a 1 element list.

Reviewer: I believe the point of this part is to show all ways to extract actual data. I am thinking that maybe we can leave this part exactly as it was, with only XPath and CSS, and we can later on add JMESPath if there is some other method to feature here.

    >>> selector.xpath('//h1')  # the same, but now with XPath
    [<Selector xpath='//h1' data='<h1>Hello, Parsel!</h1>'>]

For JSON format text, use `JMESPath`_ expressions to select fields inside the text::

    >>> text = '{"user":[{"name":"A","age":"25"},{"name":"B","age":"32"},{"name":"C","age":"19"}]}'
    >>> jmespath_selector = Selector(text=text)
    >>> jmespath_selector.jmespath('user[0]')

>     [<Selector jmespath='user[0]' data={'name': 'A', 'age': '25'}>]

Reviewer: +1 to the general approach.

In line with this, I suggest changing the preceding paragraphs to make things less HTML/XML-inclined:
- Modify the first paragraph not to mention HTML or XML: the HTML or XML text → the input text
- Move the first code snippet into the second.
- Use the same structure to introduce the HTML/XML paragraph: For HTML or XML, use …
- Change variable names: text → html, selector → html_selector, text → json, jmespath_selector → json_selector

It may also make sense to make the JSON example more minimal. The point of this section is just to show that it works, ```{"title": "Hello, Parsel!"}``` as input and `title` as selector would be more in line with the HTML example.

And extract data from those elements::

    >>> selector.css('h1::text').get()
    'Hello, Parsel!'
    >>> selector.xpath('//h1/text()').getall()
    ['Hello, Parsel!']
    >>> jmespath_selector.jmespath('user[0]').get()
    {'name': 'A', 'age': '25'}
     >>> jmespath_selector.jmespath("user[*].age.to_number(@)").getall()

>     [25, 32, 19]

Reviewer: A single, simple example should be enough here.

.. _CSS: https://www.w3.org/TR/selectors
.. _XPath: https://www.w3.org/TR/xpath
.. _JMESPath: https://jmespath.org/

Learning CSS, XPath, and JMESPath

> =================================

Reviewer: It may be best to rename this to something like “Learning expression languages”.

-   `XPath cheatsheet`_

For HTML and XML input, you can use either CSS_ or XPath_. CSS_ is usually more
readable, but some things can only be done with XPath_.

`JMESPath`_ allows you to declaratively specify how to extract elements from a JSON document.

> Resources to learn JMESPATH_ include:

Reviewer: Line length

You can use either CSS_ or XPath_. CSS_ is usually more readable, but some

> things can only be done with XPath_.

Reviewer: ```suggestion
For HTML and XML input, you can use either CSS_ or XPath_. CSS_ is usually more 
readable, but some things can only be done with XPath_.
```

     'image4_thumb.jpg',
     'image5_thumb.jpg']

JMESPath Selector
-----------------

> Whenever the selected elements possess JSON format, it is possible to use  ``.jmespath()`` method from

Reviewer: Instead of “selected elements” (“element” here can be easily confused with “HTML/XML element”), I think we should talk of “selector text” (the string the Selector selects from).

“possess” suggests having JSON (i.e. possibly only partially), “is” is probably a better choice.

```suggestion
Whenever the selector text is JSON, it is possible to use the ``.jmespath()`` method from
```

:class:`~parsel.selector.Selector`.

For example, given the selector is parsing the following element::

    >>> text = '{"user":[{"name":"A","age":"25"},{"name":"B","age":"32"},{"name":"C","age":"19"}]}'
    >>> jmespath_selector = Selector(text=text)

It is possible to extract the age of the users as an integer by using the following line of code::

    >>> jmespath_selector.jmespath("user[*].age.to_number(@)").getall()
    [25,
     32,

>      19]

Reviewer: ```suggestion
    [25, 32, 19]
```

JMESPath selectors can be chained with css and xpath selectors,

> as well as use `.re()` and `re_first()` and the other methods.

Reviewer: ```suggestion
JMESPath selectors can be chained with CSS and XPath selectors,
and they also allow you to use `.re()`, `re_first()`, and the other methods.
```

Author: I removed this part of the doc since it creates the idea that they are different. Yes, they are, but we don't need to explicitly mention it in this way. Seems more appropriate to simply present them as selectors and the user will understand they can be chained and all have the same methods.

> To find out more about JMESPath, head to `JMESPath Specification`_.

Reviewer: ```suggestion
To find out more about JMESPath, head to the `JMESPath Specification`_.
```

> .. _JMESPath Specification: https://jmespath.org/specification.html

Reviewer: Up until this section, by reading the documentation, users will have no idea of the fact that Parsel supports JMESPath selection for JSON data.

I think we need to review the documentation from the beginning, and make sure that references to JMESPath are introduced as if JMESPath support had been part of Parsel forever.

Author: I think this raises another scope for the documentation. It's true JMES can be presented in the current doc in the introductory section, I will fix that. Nevertheless, it may be good to make this redesign of the docs a task by itself maybe?

Reviewer: I have just had a proper read of the documentation, and I see what you mean that is may not be that simple.

I agree with your initial suggestion now: let this pull request be limited to this new section _and_ updating the initial README paragraph to mention JSON and JMESPath, and let the general documentation refactoring by a separate task to be done later. I actually would like to be the one to work on that.

Author: @Gallaecio I removed the JMESPath section altogether and put snippets of it in the main introductory sections. This way, although they operate differently, they stay together as much as possible.

The one thing bugging me is the necessity to create the jmespath_selector. Without it, all other simplified examples would need a bigger text base with json inside for the jmes, and also it would require some quick extractions before getting to the jmespath on every exemple, using the css or xpath.
